### PR TITLE
[FEATURE] Restore Quantization API to MXNet

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
This PR restores quantization API and some examples to master branch of MXNet. Change prepared together with @sfraczek and @grygielski 
Quantization API now utilizes HybridBlock as symbol executor and new features like `optimize_for`
Moreover:
- enabled numpy semantics support
- improved custom calibration flow (user can now use layer names which need calibration in their layer output collectors)
- renamed quantize_net_v2 to quantize_net
- using DataLoader instead of DataIter in model calibration
- retained quantize_model and quantize_model_mkldnn which works with symbol

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

@anko-intel @sfraczek @grygielski @TaoLv @pengzhao-intel @ciyongch

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

File: example/quantization/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+<!--- Unless required by applicable law or agreed to in writing, -->
+<!--- software distributed under the License is distributed on an -->
+<!--- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY -->
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# Model Quantization with Calibration Examples
+
+This folder contains examples of quantizing a FP32 model with Intel® MKL-DNN to (U)INT8 model.
```
Comment (ciyongch) [inline example/quantization/README.md:20]: As `MKL-DNN` is already renamed to `oneDNN`, it would be better to use oneDNN in the documents, what do you think?

File: example/quantization/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+example: resnet int8 performance benchmark on c5.24xlarge(duo sockets, 24 physical cores per socket).
+
+    bash ./launch_inference_mkldnn.sh -s ./model/resnet50_v1-quantized-5batches-naive-symbol.json
+
+will launch two instances for throughput benchmark and each instance will use 24 physical cores.
+```
+
+
+<h3 id='3'>ResNetV1</h3>
```
Comment (ciyongch) [inline example/quantization/README.md:114]: Previously, we listed all the models that supports quantization for better reference, is there any limitation to support other models or any other reasons?

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+                           # script if it fail
+
+    # download model
+    if not args.no_pretrained:
+        if logger:
+            logger.info('Get pre-trained model from Gluon-CV modelzoo.')
+            logger.info('If you want to use custom model, please set --no-pretrained.')
+        net, prefix = get_from_gluon(model_name=args.model, classes=1000, logger=logger)
+        rgb_mean = '0.485,0.456,0.406'
+        rgb_std = '0.229,0.224,0.225'
```
Comment (ciyongch) [inline example/quantization/imagenet_gen_qsym_mkldnn.py:183]: Does this `mean` and `std` the new value to gluon models? As I remember `mean` and `std` was ~100+ and ~58 previously.

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc [hunk 4]
```diff
@@ -137,17 +147,22 @@ void SgMKLDNNConvOperator::Forward(const OpContext &ctx,
...
       mkldnn_param.quantized ? inputs[idx++].data().dptr<float>()[0] : 0.0;
   float data_max =
       mkldnn_param.quantized ? inputs[idx++].data().dptr<float>()[0] : 0.0;
-  float sum_min = (mkldnn_param.with_sum && mkldnn_param.quantized)
-                      ? inputs[idx++].data().dptr<float>()[0]
-                      : 0.0;
-  float sum_max = (mkldnn_param.with_sum && mkldnn_param.quantized)
-                      ? inputs[idx++].data().dptr<float>()[0]
-                      : 0.0;
+  float sum_min = 0.0f;
```
Comment (ciyongch) [inline src/operator/subgraph/mkldnn/mkldnn_conv.cc:155]: Seems it's a new feature compared to v1.x, is there any test cases to cover it?

**Review by ciyongch — COMMENTED** [review]
Great work for bringing back quantization functionality to MX2.0 @bgawrych !
The PR overall looks good, I left a few comments and questions: is the model coverage same as v1.x version?

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc (hunk 4 above)
Comment (bgawrych) [reply src/operator/subgraph/mkldnn/mkldnn_conv.cc:155]: Subgraph API provide deduplication of inputs to subgraph (e.g. residual block in mobilenet)
Test for this is in tests/python/mkl/test_subgraph.py::test_dedup

File: example/quantization/README.md (hunk 1 above)
Comment (bgawrych) [reply example/quantization/README.md]: Yes! :) There is some discussion about renaming MKL-DNN to oneDNN in this RFC: https://github.com/apache/incubator-mxnet/issues/19610
For now I've changed only occurrences in example/quantization/* and used full name in this particular line

File: example/quantization/README.md (hunk 2 above)
Comment (bgawrych) [reply example/quantization/README.md:114]: We choose this ResNetV1 as an example - previously in this file there was more models but content was almost the same (changed model name only).
Moreover only some models work with script, because of this issue: https://github.com/apache/incubator-mxnet/issues/19580 and GluonCV for master branch is not compatible yet
We will add references when more models will be available

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu, edge]

File: example/quantization/README.md (hunk 1 above)
Comment (ciyongch) [reply example/quantization/README.md]: That's more than what I mean, but it's good :) I was referring to only change the name from `MKL-DNN` to `oneDNN` in the document file or the relative description of the source code file at this moment :)

File: example/quantization/README.md (hunk 2 above)
Comment (ciyongch) [reply example/quantization/README.md:114]: Ok, so can we add the model list which are already supported and verified?

File: example/quantization/imagenet_inference.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    parser.add_argument('--num-warmup-batches', type=int, default=5, help='number of warmup batches used for benchmark')
+    parser.add_argument('--shuffle-dataset', action='store_true', default=True,
+                        help='shuffle the score dataset')
+    parser.add_argument('--data-layer-type', type=str, default='float32',
+                        choices=['float32', 'int8', 'uint8'],
+                        help='data type for data layer (only with --benchmark)')
+
+    args = parser.parse_args()
+
+    ctx = mx.cpu(0)
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply example/quantization/imagenet_inference.py]: I restored ctx and added warning when using GPU.

I don't think there are any limitations for bf16/fp32 - we just removed it to be consistent as example folder is called quantization. We can add bfloat examples to automatic-mixed-precision folder in other PR

File: example/quantization/README.md (hunk 2 above)
Comment (bgawrych) [reply example/quantization/README.md]: I've added verified models with tested accuracy

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu]

File: example/quantization/imagenet_inference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ciyongch) [reply example/quantization/imagenet_inference.py]: Sure, sounds good :)

**Review by ciyongch — APPROVED** [review]
LGTM

@TaoLv @xinyu-intel @pengzhao-intel , any other comments?

File: example/quantization/imagenet_gen_qsym_onednn.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    return net, prefix
+
+def regex_find_excluded_symbols(patterns_dict, model_name):
+    for key, value in patterns_dict.items():
+        if re.search(key, model_name) is not None:
+            return value
+    return None
+
+def get_exclude_symbols(model_name, exclude_first_conv):
+    # Grouped supported models at the time of commit:
```
Comment (TaoLv) [inline example/quantization/imagenet_gen_qsym_onednn.py:57]: Better to use doc string?

File: example/quantization/imagenet_gen_qsym_onednn.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            'vgg': ['vgg_hybridsequential0_conv2d0_fwd'],
+        }
+        excluded_first_conv_sym_names = regex_find_excluded_symbols(first_conv_regex, model_name)
+        if excluded_first_conv_sym_names is None:
+            raise ValueError('Currently, model %s is not supported in this script' % model_name)
+        excluded_sym_names += excluded_first_conv_sym_names
+    return excluded_sym_names
+
+
+
```
Comment (TaoLv) [inline example/quantization/imagenet_gen_qsym_onednn.py:97]: It looks like too many blank lines.

**Comment by TaoLv** [comment]
Do we need to start a new RFC or design doc as demonstrated originally in [1] and [2]?

[1] https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN
[2] https://github.com/apache/incubator-mxnet/pull/9552

File: example/quantization/README.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# Model Quantization with Calibration Examples
+
+This folder contains examples of quantizing a FP32 model with Intel® oneAPI Deep Neural Network Library (oneDNN) to (U)INT8 model.
+
+<h2 id="0">Contents</h2>
+
+* [1. Model Quantization with Intel® oneDNN](#1)
```
Comment (xinyu-intel) [inline example/quantization/README.md:24]: since cudnn quantization will not be re-enabled and introduced in this README, so we can just remove this contents.

**Review by xinyu-intel — APPROVED** [review]
LGTM:)

File: example/quantization/README.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+<h2 id="1">Model Quantization with Intel® oneDNN</h2>
+
+Intel® oneDNN supports quantization with subgraph features on Intel® CPU Platform and can bring performance improvements on the [Intel® Xeon® Scalable Platform](https://www.intel.com/content/www/us/en/processors/xeon/scalable/xeon-scalable-platform.html). To apply quantization flow to your project directly, please refer [Optimize custom models with oneDNN backend](#TODO(agrygielski)).
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply example/quantization/README.md]: Removed - we will add link to article once it will be published

File: example/quantization/README.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                        quantized models if the calibration dataset is
+                        representative enough of the inference dataset.
+                        default is `entropy`
+  --quantized-dtype {auto,int8,uint8}
+                        quantization destination data type for input data,
+                        default is `auto`
+  --quiet               suppress most of log
+```
+
+A new benchmark script `launch_inference_onednn.sh` has been designed to launch performance benchmark for float32 or int8 image-classification models with Intel® oneDNN.
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply example/quantization/README.md]: Done

File: example/quantization/imagenet_gen_qsym_onednn.py (hunk 6 above)
Comment (bgawrych) [reply example/quantization/imagenet_gen_qsym_onednn.py]: Done

File: example/quantization/imagenet_gen_qsym_onednn.py (hunk 7 above)
Comment (bgawrych) [reply example/quantization/imagenet_gen_qsym_onednn.py]: Done

File: example/quantization/README.md (hunk 8 above)
Comment (bgawrych) [reply example/quantization/README.md]: Right, done

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by bgawrych** [comment]
> Do we need to start a new RFC or design doc as demonstrated originally in [1] and [2]?
> 
> [1] https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN
> [2] #9552

We haven't changed design of quantization - all code changes in backend are related to changed node naming conventions. I don't know if in this case we must propose new RFC as previous one was accepted and all what we have done here is bringing it back

**Comment by pengzhao-intel** [comment]
We don't need a new RFC since this is the same approach (quantization flow) migration from 1.x to 2.x.

**Review by pengzhao-intel — APPROVED** [review]
LGTM and I will merge soon if no other comments.

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by bgawrych** [comment]
@Szha is something wrong with unix-gpu CI? I got following error two times in a row:
- [2020-12-10T12:17:34.769Z] Cannot contact mxnetlinux-cpu_cwcmun1huz: java.lang.InterruptedException
...
[2020-12-10T12:28:54.084Z] FAILED: CMakeFiles/mxnet.dir/src/operator/numpy/np_kron.cc.o 
...
[2020-12-10T12:28:54.084Z] c++: fatal error: Killed signal terminated program cc1plus

but in run before in GPU: MKLDNN job it was CMakeFiles/mxnet.dir/src/operator/numpy/np_elemwise_broadcast_logic_op.cc.o 

I don't think it's releated to my change

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by leezu** [comment]
@bgawrych you can refer to https://github.com/apache/incubator-mxnet/issues/19623 for the CI issue. It looks like an effect of https://github.com/apache/incubator-mxnet/issues/18501 together with memory usage increase in recent versions of gcc. If you have time to help fix it, that would be great.

**Comment by pengzhao-intel** [comment]
It's great the CI passed. 

Thanks the great efforts from the team to re-enable quantization flow in the MXNet 2.0: )

I am going to merge this PR. @anko-intel @bgawrych

**Review by pengzhao-intel — APPROVED** [review]
Merging now.
